### PR TITLE
feat: bitblasting theorems for signed comparisons

### PR DESCRIPTION
**Pull request by alexkeizer**
Prove theorems that relate `BitVec.slt` and `sle` to `carry`, so that these signed comparisons may be bitblasted in LeanSAT.

This PR is stacked on top of #4200. For the diff without changes from that PR, see: https://github.com/opencompl/lean4/compare/opencompl:lean4:bitvec-toInt-iff-msb...bitvec-slt-blast

### CODE REVIEW

File: src/Init/Data/BitVec/Bitblast.lean [hunk 1]
```diff
@@ -198,4 +198,53 @@ theorem ule_eq_not_ult (x y : BitVec w) : x.ule y = !y.ult x := by
 theorem ule_eq_carry (x y : BitVec w) : x.ule y = carry w y (~~~x) true := by
   simp [ule_eq_not_ult, ult_eq_not_carry]
 
+/-- If two bitvectors have the same `msb`, then signed and unsigned comparisons coincide -/
+theorem slt_eq_ult_of_msb_eq {x y : BitVec w} (h : x.msb = y.msb) :
+    x.slt y = x.ult y := by
+  simp only [BitVec.slt, toInt_eq_msb_cond, BitVec.ult, decide_eq_decide, h]
+  cases y.msb <;> simp only [↓reduceIte, Int.ofNat_lt, Int.sub_lt_sub_right_iff]
```
Comment (semorrison) [inline src/Init/Data/BitVec/Bitblast.lean:205]: I think we can just make `Int.sub_lt_sub_left_iff` (and `right`) `@[simp]` lemmas, and then just 
```suggestion
  simp only [BitVec.slt, toInt_eq_msb_cond, BitVec.ult, h]
  cases y.msb <;> simp
```

File: src/Init/Data/BitVec/Bitblast.lean [hunk 2]
```diff
@@ -198,4 +198,53 @@ theorem ule_eq_not_ult (x y : BitVec w) : x.ule y = !y.ult x := by
...
+
+/-- If two bitvectors have different `msb`s, then unsigned comparison is determined by this bit -/
+theorem ult_eq_msb_of_msb_neq {x y : BitVec w} (h : x.msb ≠ y.msb) :
+    x.ult y = y.msb := by
+  simp [BitVec.ult]
```
Comment (semorrison) [inline src/Init/Data/BitVec/Bitblast.lean:210]:
```suggestion
  simp only [BitVec.ult]
```

File: src/Init/Data/BitVec/Bitblast.lean [hunk 3]
```diff
@@ -198,4 +198,53 @@ theorem ule_eq_not_ult (x y : BitVec w) : x.ule y = !y.ult x := by
...
+  cases hy : y.msb <;> simp only [decide_eq_true_eq, decide_eq_false_iff_not, Nat.not_lt, ge_iff_le]
```
Comment (semorrison) [inline src/Init/Data/BitVec/Bitblast.lean:211]:
```suggestion
  cases hy : y.msb <;> simp only [decide_eq_true_eq, decide_eq_false_iff_not]
```

File: src/Init/Data/BitVec/Bitblast.lean [hunk 4]
```diff
@@ -198,4 +198,53 @@ theorem ule_eq_not_ult (x y : BitVec w) : x.ule y = !y.ult x := by
...
+  · have hx : x.msb = true := by simp_all
+    have := (msb_eq_false_iff_two_mul_lt _).mp hy
+    have := (msb_eq_true_iff_two_mul_ge _).mp hx
+    omega
+  · have hx : x.msb = false := by simp_all
+    have := (msb_eq_false_iff_two_mul_lt _).mp hx
+    have := (msb_eq_true_iff_two_mul_ge _).mp hy
+    omega
```
Comment (semorrison) [inline src/Init/Data/BitVec/Bitblast.lean:219]:
```suggestion
  simp only [BitVec.ult, msb_eq_decide, ne_eq, decide_eq_decide] at *
  omega
```

File: src/Init/Data/BitVec/Bitblast.lean [hunk 5]
```diff
@@ -198,4 +198,53 @@ theorem ule_eq_not_ult (x y : BitVec w) : x.ule y = !y.ult x := by
...
+
+/-- If two bitvectors have different `msb`s, then signed and unsigned comparisons are opposites -/
+theorem slt_eq_not_ult_of_msb_neq {x y : BitVec w} (h : x.msb ≠ y.msb) :
+    x.slt y = !x.ult y := by
+  have hx : x.msb = !y.msb := by revert h; cases x.msb <;> cases y.msb <;> trivial
```
Comment (semorrison) [inline src/Init/Data/BitVec/Bitblast.lean:224]: Sadly this lemma doesn't seem to exist, but perhaps it should: `Bool.eq_not_of_ne`.

File: src/Init/Data/BitVec/Bitblast.lean [hunk 6]
```diff
@@ -198,4 +198,53 @@ theorem ule_eq_not_ult (x y : BitVec w) : x.ule y = !y.ult x := by
...
+  simp only [BitVec.slt, toInt_eq_msb_cond, hx, not_eq_true', ult_eq_msb_of_msb_neq h]
+  cases y.msb
+  <;> simp only [↓reduceIte, false_eq_true, Bool.not_false, decide_eq_true_eq, gt_iff_lt,
+        Bool.not_true, decide_eq_false_iff_not]
+  <;> omega
```
Comment (semorrison) [inline src/Init/Data/BitVec/Bitblast.lean:229]: I would be satisfied with
```
  simp only [BitVec.slt, toInt_eq_msb_cond, hx, ult_eq_msb_of_msb_neq h]
  cases y.msb <;> (simp; omega)
```
here (i.e. no need to squeeze the simp if followed by `omega`).

**Comment by semorrison** [comment]
Otherwise, LGTM.

File: src/Init/Data/BitVec/Bitblast.lean (hunk 5 above)
Comment (alexkeizer) [reply src/Init/Data/BitVec/Bitblast.lean]: I've added `Bool.eq_not_of_ne`

**Comment by alexkeizer** [comment]
@semorrison I've addressed all your comments. As soon as #4200 is merged, we can rebase this PR and it should be ready to go, too

**Comment by alexkeizer** [comment]
@semorrison I've rebased, this should be ready to merge

**Pull request merged**